libs/platform/ebpf_epoch.c

```diff
@@ -589,140 +608,377 @@ _ebpf_epoch_release_free_list(_Inout_ ebpf_epoch_cpu_entry_t* cpu_entry, int64_t
-            if (_ebpf_epoch_cpu_table[cpu_id].stale) {
-                if (!_ebpf_epoch_cpu_table[cpu_id].stale_worker) {
-                    // If ebpf_allocate_non_preemptible_work_item fails, it will retry next time the timer fires.
-                    (void)ebpf_allocate_non_preemptible_work_item(
-                        &_ebpf_epoch_cpu_table[cpu_id].stale_worker, cpu_id, _ebpf_epoch_stale_worker, NULL);
-                }
-                if (_ebpf_epoch_cpu_table[cpu_id].stale_worker) {
-                    KeInsertQueueDpc(_ebpf_epoch_cpu_table[cpu_id].stale_worker, NULL, NULL);
-                }
-            } else {
-                _ebpf_epoch_cpu_table[cpu_id].stale = true;
-            }
-        }
+    if (_ebpf_epoch_cpu_table[0].rundown_in_progress) {
+        return;
+    }
 
-        // Loop over the thread table and compute the lowest active epoch.
-        for (size_t index = 0; index < EBPF_EPOCH_THREAD_TABLE_SIZE; index++) {
-            ebpf_epoch_state_t* epoch_state = &_ebpf_epoch_cpu_table[cpu_id].epoch_table[index];
-            if (epoch_state->epoch != 0) {
-                lowest_epoch = min(lowest_epoch, epoch_state->epoch);
-            }
-        }
+    if (!_ebpf_epoch_cpu_table[0].epoch_computation_in_progress) {
+        _ebpf_epoch_cpu_table[0].epoch_computation_in_progress = true;
+        _ebpf_epoch_skipped_timers = 0;
+        _ebpf_epoch_compute_release_epoch_message.message_type = EBPF_EPOCH_CPU_MESSAGE_TYPE_COMPUTE_EPOCH;
+        MemoryBarrier();
+        _ebpf_epoch_compute_release_epoch_message.message.compute_epoch.epoch = _ebpf_current_epoch;
+        KeInitializeEvent(&_ebpf_epoch_compute_release_epoch_message.completion_event, NotificationEvent, false);
+        KeSetTargetProcessorDpc(&_ebpf_epoch_compute_release_epoch_dpc, 0);
+        KeInsertQueueDpc(&_ebpf_epoch_compute_release_epoch_dpc, &_ebpf_epoch_compute_release_epoch_message, NULL);
+    } else {
+        _ebpf_epoch_skipped_timers++;
+        LARGE_INTEGER due_time;
+        due_time.QuadPart = -(EBPF_EPOCH_FLUSH_DELAY_IN_NANOSECONDS / EBPF_NANO_SECONDS_PER_FILETIME_TICK);
+        KeSetTimer(&_ebpf_epoch_compute_release_epoch_timer, due_time, &_ebpf_epoch_timer_dpc);
+    }
+}
+
+typedef bool (*ebpf_epoch_messenger_worker_t)(
+    _In_ KDPC* dpc,
+    _Inout_ ebpf_epoch_cpu_entry_t* cpu_entry,
+    _Inout_ ebpf_epoch_cpu_message_t* message,
+    uint32_t current_cpu,
+    _Out_ uint32_t* next_cpu);
+
+/**
+ * @brief Compute the lowest epoch on this CPU. If this is the last CPU, then set the release epoch and start re-arming
+ * the timer.
+ *
+ * @param[in] dpc DPC that triggered this function.
+ * @param[in] cpu_entry CPU entry to compute the epoch for.
+ * @param[in] message Message to process.
+ * @param[in] current_cpu Current CPU.
+ * @param[out] next_cpu Next CPU to send the message to.
+ * @retval true Forward the message to the next CPU.
+ * @retval false Do not forward the message to the next CPU.
+ *
```

Comment:
can you please add the message_id that triggers this worker?

Reply:
Fixed.